catroid/src/androidTestCreateatschool/java/org/catrobat/catroid/createatschool/uiespresso/ui/activity/RestrictedLoginTest.java

```diff
@@ -0,0 +1,232 @@
+import org.catrobat.catroid.uiespresso.util.BaseActivityInstrumentationRule;
+import org.catrobat.catroid.uitest.util.UiTestUtils;
+import org.catrobat.catroid.web.ServerCalls;
+import org.hamcrest.Matcher;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+import static android.support.test.espresso.Espresso.onView;
+import static android.support.test.espresso.Espresso.openActionBarOverflowOrOptionsMenu;
+import static android.support.test.espresso.Espresso.pressBack;
+import static android.support.test.espresso.action.ViewActions.clearText;
+import static android.support.test.espresso.action.ViewActions.click;
+import static android.support.test.espresso.action.ViewActions.typeText;
+import static android.support.test.espresso.assertion.ViewAssertions.doesNotExist;
+import static android.support.test.espresso.assertion.ViewAssertions.matches;
+import static android.support.test.espresso.intent.Intents.intended;
+import static android.support.test.espresso.intent.Intents.intending;
+import static android.support.test.espresso.intent.matcher.IntentMatchers.hasAction;
+import static android.support.test.espresso.intent.matcher.IntentMatchers.hasData;
+import static android.support.test.espresso.matcher.ViewMatchers.isDisplayed;
+import static android.support.test.espresso.matcher.ViewMatchers.withId;
+import static android.support.test.espresso.matcher.ViewMatchers.withResourceName;
+import static android.support.test.espresso.matcher.ViewMatchers.withText;
+
+import static junit.framework.Assert.assertEquals;
+
+import static org.hamcrest.Matchers.allOf;
+
+@RunWith(AndroidJUnit4.class)
+public class RestrictedLoginTest {
+
+	@Rule
+	public BaseActivityInstrumentationRule<CreateAtSchoolMainMenuActivity> baseActivityTestRule = new
+			BaseActivityInstrumentationRule<>(CreateAtSchoolMainMenuActivity.class, true, false);
+
+	private List<String> allSettings = new ArrayList<>(Arrays.asList(Constants.RESTRICTED_LOGIN_ACCEPTED, Constants.CREATE_AT_SCHOOL_USER));
+	private Map<String, Boolean> initialSettings = new HashMap<>();
+	private IdlingResource idlingResource;
+
+	@Before
+	public void setUp() {
+		SharedPreferences sharedPreferences = PreferenceManager
+				.getDefaultSharedPreferences(InstrumentationRegistry.getTargetContext());
+
+		for (String setting : allSettings) {
+			initialSettings.put(setting, sharedPreferences.getBoolean(setting, false));
+		}
+
+		SharedPreferences.Editor editor = sharedPreferences.edit();
+		editor.putBoolean(Constants.RESTRICTED_LOGIN_ACCEPTED, false);
+		editor.putBoolean(Constants.CREATE_AT_SCHOOL_USER, true);
+		editor.apply();
+
+		Reflection.setPrivateField(Constants.class, "DEFAULT_ROOT", Environment.getExternalStorageDirectory()
```

Comment:
why does this have to run in the default root again? if thats really necessary, theres probably nicer ways to achieve that.

the default root is reflected out (the only reflection that is currently in place in the whole espresso setup) in the BaseActivityInstrumentationRule to be "pocket code ui test". 
i did this because in the old setup, the setup went to the project folder and zipped and deleted all projects, and then in teardown unzipped everything again and deleted the zip.
i thought this isnt actually that nice because it leads to a lot of storage operations and zipping on every single test, also if the instrumenter crashes, your projects probably going to stay zipped in your folder, etc...

so i switched to reflecting the DEFAULT_ROOT to "pocket code ui test" folder, so that every test just deletes this before the test runs. which has the nice effect also that if you run a single test, you can inspect the project xmls, as they are still in place after the test has finished, so only the next test would clean that up in startup.

if for some reason you do really really need to use the production "pocket code" folder, you should write your own ActivityInstrumentationRule that does not do the reflection and deals with existing projects itself.

so in short:
as you do here is not nice, first, because you could just get rid of two reflections by making your own ActivityInstrumentationRule, and second, which also applies if you do your own rule, you have to deal with existing projects and to cleanup. (keep in mind cornercases where you already have a project with exactly this name, etc...)
so if theres no real good reason to do the reflection here, just keep the "pocket code uitests" one and work around it another way.

Reply:
I had some troubles executing the tests in bulk, when there's no project - so with an empty "pocket code uitests" folder. There have been some strange exceptions going on in MainMenuActivitiy's onResume-method (UtilFile.createStandardProjectIfRootDirectoryIsEmpty), which seemed to me a bit flaky due to the different folder. However, I couldn't reproduce it now. Neither on the phone nor on the emulator. So, maybe there was some other issue going on. I've tried it several times again, it's working fine, so I've removed that line again.

Reply:
regarding strange exceptions in main menu activity.

this could also arise from the activity switching (as already mentioned a few times). i found that sometimes this causes tests to hang / crash, since for some f*cked up reason a new test starting the main menu activity ends up in another activity.

reproducable example:
start the script activity via test rule
add a new scene via the switch scene brick
automatically puts you into scene list
end test

-> next test starting the main menu activity will break for sure (can be a lot of tests later also), because for some reason it ends up int eh program activity

not so reproducable example:
starting an activity with the test rule that starts a background thread and close it quick again (eg. short test)
can lead to either activity not being killed while next test is starting and the times out because hes waiting for a activity that the rule is never going to start. while ive observed this only happens when running them in bulk, if running standalone activity is finished.

so a lot of weird stuff happens due to background threads and activity switches. a lot of those problems are probably due to us messing around with the back-stack of the activities, and with not-so-android-best-practice way of starting / finishing activities and doing stuff in activity lifecyle methods that shouldnt be done at those points in time.

anyways, what i ended up doing to locate tests that do cause problems with other tests:
1. write a testsuite with all your tests (so youve got control over what test classes are run and in which order)
2. check that the problem now occurs
3. check that the flaky test works when run alone
4. start commenting out blocks of testclasses and run testsuite to see which is affecting the test

ive done this now a few times, and its a pain. so this is why i recommend limiting switching inbetween activities as far as possible. (especially main menu activity and other that use background threads proofed to be problematic)

in core espresso test suite ive got 180 tests of which ~140 are run with the pullrequest trigger, and currently those are stable and not affecting each other, many use the main menu activity. that took my countless hours trying to figure out what causes flakyness, learn from my fails...
there are also certain ui elements that cause flakyness out of the box, like spinners, googling that there seems to be a LOT of people having problems with those, so this is why i introduced the @Flaky annotation that will run test up to 3 times (default, can be changed with eg @Flaky(5)) before showing it as failed test. (but if theres tests affecting each other in above described way, no avail with @Flaky...)

Reply:
Ok, I understand .. this really sounds like a pain.
Finally, I could get rid of activity restarting in this test class, so it should be fine now.